test/jdk/java/net/httpclient/HttpResponseConnectionLabelTest.java

```diff
@@ -0,0 +1,446 @@
+        LOGGER.log("Firing request 2...");
+        CompletableFuture<HttpResponse<String>> response2Future =
+                client.sendAsync(pair2.request, BodyHandlers.ofString(CHARSET));
+
+        // Release latches to allow the server handlers to proceed
+        latch.countDown();
+        latch.countDown();
+
+        // Wait for all parties to be ready
+        LOGGER.log("Client is waiting for the latch...");
+        latch.await();
+        LOGGER.log("Client is continuing...");
+
+        // Collect responses
+        HttpResponse<String> response1 = response1Future.get();
+        HttpResponse<String> response2 = response2Future.get();
+
+        // Verify successful responses
+        assertEquals(200, response1.statusCode());
+        assertEquals(200, response2.statusCode());
+
+        // Verify that connection keys differ; that is, requests are served through different connections
+        String connectionKey1 = response1.headers().firstValue(CONNECTION_KEY_HEADER_NAME).get();
+        String connectionKey2 = response2.headers().firstValue(CONNECTION_KEY_HEADER_NAME).get();
+        assertNotEquals(connectionKey1, connectionKey2);
+
+        // Verify that server IDs differ.
+        // (Using `parseInt` to validate the content.)
+        int serverId1 = response1.headers().firstValue(SERVER_ID_HEADER_NAME).map(Integer::parseInt).get();
+        int serverId2 = response2.headers().firstValue(SERVER_ID_HEADER_NAME).map(Integer::parseInt).get();
+        assertNotEquals(serverId1, serverId2);
+
+        // Verify that response bodies differ.
+        // (Using `parseInt` to validate the content.)
+        int body1 = Integer.parseInt(response1.body());
+        int body2 = Integer.parseInt(response2.body());
+        assertNotEquals(body1, body2);
+
+        // Verify that connection labels differ; that is, requests are served through different connections
+        String label1 = response1.connectionLabel().orElse(null);
+        assertNotNull(label1);
+        LOGGER.log("connection label 1: %s", label1);
+        String label2 = response2.connectionLabel().orElse(null);
+        assertNotNull(label2);
+        LOGGER.log("connection label 2: %s", label2);
+        assertNotEquals(label1, label2);
+
+    }
+
+    static Stream<ServerRequestPair> testSerialRequestsToSameServer() {
+        return Stream.of(PRI_HTTP1, PRI_HTTPS1, PRI_HTTP2, PRI_HTTPS2);
+    }
+
+    @ParameterizedTest
+    @MethodSource("testSerialRequestsToSameServer")
+    void testSerialRequestsToSameServer(ServerRequestPair pair) throws Exception {
+
+        // Disarm the synchronization point
+        pair.serverResponseLatchRef[0] = null;
+
+        // Fire requests
+        LOGGER.log("Firing request 1...");
+        HttpResponse<String> response1 = client.send(pair.request, BodyHandlers.ofString(CHARSET));
+        LOGGER.log("Firing request 2...");
+        HttpResponse<String> response2 = client.send(pair.request, BodyHandlers.ofString(CHARSET));
```

Comment:
In theory, there's no guarantee that these sequential requests will be executed "immediately" one after the other. Internally, in the httpclient implementation, we use idle timeouts to close idle connections. So I'm wondering if it's realistic that there would ever be a case where in some setup (like the CI), these two execute so far apart from each other (`-Xcomp`?) that the connection might have timed out in the meantime and closed? Thus the second request ends up using a different connection and fails this test?
Should we perhaps use `othervm` for this test to configure an extremely high idle timeout of connections to avoid such intermittent failures?

Reply:
Waiting for the test to fail due to this reason and then adjusting the timeout is fine with me.

Reply:
On the plus side - if we add the `-Djdk.httpclient.keepalive.timeout=120` now we might not have to deal with it later, and maybe it will give us a clue if the test ever fails due to this.

Reply:
Added `-Djdk.httpclient.keepalive.timeout=120` in edc0c8d84e68505d054fa87b24316f456b15c4f0.